### PR TITLE
User community wants RTPS statistics

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

The user community want the ability to query OpenDDS for statistics
related to RTPS.  Specifically, the number of messages send/received
per address and counts related to retry.

Solution
--------

Build on the existing RtpsRelay counts.

### CODE REVIEW

File: java/dds/.gitignore [hunk 1]
```diff
@@ -124,39 +123,42 @@
...
+/OpenDDS_DCPS_transport_MulticastInst.h
+/OpenDDS_DCPS_transport_RtpsUdpInst.h
+/OpenDDS_DCPS_transport_TcpInst.h
+/OpenDDS_DCPS_transport_TheTransportRegistry.h
+/OpenDDS_DCPS_transport_TransportConfig.h
+/OpenDDS_DCPS_transport_TransportInst.h
+/OpenDDS_DCPS_transport_UdpInst.h
+/OpenddsDcpsExtJC.cpp
+/OpenddsDcpsExtJC.h
+/classes
```
Comment (mitza-oci) [inline java/dds/.gitignore]: newline needed

File: dds/OpenddsDcpsExt.idl [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    };
+
+    typedef sequence<Locator_t> LocatorSeq;
+
+    typedef long MessageCountKind;
+    const MessageCountKind MCK_RTPS = 0;
+    const MessageCountKind MCK_STUN = 1;
+
+    struct MessageCount {
+      @key Locator_t locator;
```
Comment (mitza-oci) [inline dds/OpenddsDcpsExt.idl:98]: Should this use `BUILT_IN_TOPIC_KEY`?

File: dds/DCPS/transport/framework/NetworkAddress.h [hunk 3]
```diff
@@ -95,6 +96,44 @@ ACE_INET_Addr choose_single_coherent_address(const ACE_INET_Addr& addr, bool pre
 extern OpenDDS_Dcps_Export
 ACE_INET_Addr choose_single_coherent_address(const String& hostname, bool prefer_loopback = true, bool allow_ipv4_fallback = true);
 
+inline void assign(DDS::OctetArray16& dest,
+                   ACE_CDR::ULong ipv4addr_be)
+{
+  std::memset(&dest[0], 0, 12);
```
Comment (mitza-oci) [inline dds/DCPS/transport/framework/NetworkAddress.h:104]: #include <cstring>

File: dds/DCPS/transport/framework/NetworkAddress.cpp [hunk 4]
```diff
@@ -808,6 +808,74 @@ ACE_INET_Addr choose_single_coherent_address(const String& address, bool prefer_
...
+      locator.kind = LOCATOR_KIND_UDPv4;
+      const sockaddr_in* in = static_cast<const sockaddr_in*>(raw);
+      std::memset(&locator.address[0], 0, 12);
+      std::memcpy(&locator.address[12], &in->sin_addr, 4);
+    }
+    break;
+#ifdef ACE_HAS_IPV6
+  case AF_INET6:
+    {
+      locator.kind = RTPS::LOCATOR_KIND_UDPv6;
```
Comment (mitza-oci) [inline dds/DCPS/transport/framework/NetworkAddress.cpp:865]:
```suggestion
      locator.kind = LOCATOR_KIND_UDPv6;
```

File: dds/DCPS/transport/framework/TransportStatistics.h [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+struct InternalTransportStatistics {
+  const OPENDDS_STRING transport;
+  typedef OPENDDS_MAP(InternalMessageCountKey, InternalMessageCountValue) MessageCountMap;
+  MessageCountMap message_count;
+  typedef OPENDDS_MAP(GUID_t, CORBA::ULong) GuidCountMap;
+  GuidCountMap writer_resend_count;
+  GuidCountMap reader_nack_count;
+
+  InternalTransportStatistics(const OPENDDS_STRING& a_transport)
```
Comment (mitza-oci) [inline dds/DCPS/transport/framework/TransportStatistics.h:94]:
```suggestion
  explicit InternalTransportStatistics(const OPENDDS_STRING& a_transport)
```

File: dds/DCPS/transport/framework/TransportStatistics.h [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  {
+    message_count.clear();
+    writer_resend_count.clear();
+    reader_nack_count.clear();
+  }
+};
+
+inline void append(TransportStatisticsSequence& seq, const InternalTransportStatistics& istats)
+{
+  size_t idx = seq.length();
```
Comment (mitza-oci) [inline dds/DCPS/transport/framework/TransportStatistics.h:108]:
```suggestion
  const ACE_CDR::ULong idx = seq.length();
```

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.h [hunk 7]
```diff
@@ -90,7 +91,9 @@ class OpenDDS_Rtps_Udp_Export RtpsUdpDataLink : public DataLink, public virtual
   RtpsUdpDataLink(RtpsUdpTransport& transport,
                   const GuidPrefix_t& local_prefix,
                   const RtpsUdpInst& config,
-                  const ReactorTask_rch& reactor_task);
+                  const ReactorTask_rch& reactor_task,
+                  DCPS::InternalTransportStatistics& transport_statistics,
```
Comment (mitza-oci) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.h:95]:
```suggestion
                  InternalTransportStatistics& transport_statistics,
```
probably a few more of these below

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: dds/OpenddsDcpsExt.idl (hunk 2 above)
Comment (jrw972) [reply dds/OpenddsDcpsExt.idl:98]: I don't think so.  If they ever make it into a BIT then we change it.

**Review by mitza-oci — APPROVED** [review]

File: dds/DCPS/RTPS/Sedp.h [hunk 8]
```diff
@@ -655,7 +655,7 @@ class Sedp : public DCPS::RcEventHandler {
     type_lookup_service_ = type_lookup_service;
   }
 
-  bool should_drop(ssize_t length) const;
+  RcHandle<DCPS::TransportInst> transport_inst() const { return transport_inst_; }
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Sedp.h:658]:
```suggestion
  bool should_drop(ssize_t length) const;
```

**Pull request merged**